<center><img src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRo_pfye1av2mPhdiMEDZMlSpzJPm5SIIOWrg&s" width="100"></center>

**LEMBAR KERJA 3**

**UJI HIPOTESIS 1 – ONE SAMPLE Z-TEST**

**PROGRAM STUDI S1 INFORMATIKA**

**UNIVERSITAS DIPONEGORO**

--- 

**Oleh:**

**Nama :** Muchammad Yuda Tri Ananda

**NIM :** 24060124110142

**Kelas :** D

---

Untuk bisa mengakses lembar kerja ini, buka link berikut:
https://s.id/jCUIs

**Aturan Umum:**
* Print Lembar Kerja ini
* Kerjakan menggunakan tulisan tangan untuk setiap item pertanyaan berikut ini
* Kumpulkan Lembar Kerja ini pada pertemuan selanjutnya disertai dengan pengumpulan file kode program dengan nama file “LK3_Nim_NamaDepan.ipynb” ke email dari pengampu r.kusumaningrum81@gmail.com (Dosen Pengampu: Dr. Retno Kusumaningrum, S.Si., M.Kom.) atau sandyk@lecturer.undip.ac.id (Dosen Pengampu: Sandy Kurniawan, S.Kom., M.Kom.)

**Aturan Penulisan Kode Program:**
* Gunakan Google Collaboratory untuk mengerjakan lembar kerja terkait.

## Soal 1
Buatlah data simulasi berupa 100 kali lemparan dadu yang tidak adil, dimana angka 6 memiliki kemungkinan 2x lebih besar untuk muncul pada setiap lemparan. Gunakan Z-Test untuk mengetahui apakah simulasi lemparan dadu tersebut kemungkinan besar berasal dari dadu yang adil, jika diketahui rata-rata lemparan dadu yang diharapkan adalah 3.5?

**Langkah #1:** Import Z-Test function dari statsmodel library.

In [26]:
from statsmodels.stats.weightstats import ztest
import random
import pandas as pd
import numpy as np # Berguna untuk mean, std, meskipun ztest bisa menghitungnya

**Langkah #2:** Buat simulasi data 100 kali lemparan dadu, dimana angka 6 memiliki 2x kemungkinan lebih sering muncul dibandingkan angka-angka lainnya

In [27]:
def simulasi_dadu(num_rolls):
    simulasi_list = []
    choices = [1, 2, 3, 4, 5, 6, 6] 
    for _ in range(num_rolls):
        lemparan = random.choice(choices)
        simulasi_list.append({"lemparan:": lemparan}) 
    df = pd.DataFrame(simulasi_list)
    return df

# Simulasi 100x lemparan dadu
df_100 = simulasi_dadu(100)

print("Contoh 5 data pertama dari df_100:")
print(df_100.head())
print(f"\nBentuk DataFrame df_100: {df_100.shape}")
print(f"Nama kolom: {df_100.columns}")

Contoh 5 data pertama dari df_100:
   lemparan:
0          1
1          6
2          5
3          4
4          5

Bentuk DataFrame df_100: (100, 1)
Nama kolom: Index(['lemparan:'], dtype='object')


**Jika ada yang salah dengan code tersebut, maka seharusnya:**

Kode pada Langkah #2 yang diberikan dalam soal asli:
```python
# Simulasi 100x lemparan dadu, angka 6 lebih sering muncul 2x
simulasi = []
for i in range (100):
    lemparan = random.choice ((1, 2, 3, 4, 5, 6, 6])
    d = {"lemparan:": lemparan}
    simulasi.append(d)

# Convert data simulasi ke dalam data frame
df = pd.DataFrame (simulasi)
```
**Analisis:**
Kode tersebut sudah benar dalam menciptakan simulasi di mana angka 6 memiliki kemungkinan 2x lebih besar untuk muncul dibandingkan masing-masing angka lainnya (1, 2, 3, 4, atau 5). Sample space `(1, 2, 3, 4, 5, 6, 6)` berarti setiap angka 1-5 memiliki probabilitas 1/7, dan angka 6 memiliki probabilitas 2/7.

Pembuatan DataFrame dengan `d = {"lemparan:": lemparan}` dan `simulasi.append(d)` menghasilkan DataFrame dengan nama kolom `"lemparan:"` (dengan titik dua). Ini penting untuk diingat saat memanggil kolom tersebut di langkah selanjutnya (misalnya, `df['lemparan:']`).

Fungsi `simulate_dice_rolls_prompt_format` yang saya buat di atas sudah mengimplementasikan logika yang sama dan menghasilkan format DataFrame yang serupa.

**Langkah #3:** Menampilkan hasil data simulasi

In [28]:
print("Data simulasi (100 lemparan) df_100:")
print(df_100) # Menampilkan 5 baris pertama
print(f"\nJumlah data: {len(df_100)}")

Data simulasi (100 lemparan) df_100:
    lemparan:
0           1
1           6
2           5
3           4
4           5
..        ...
95          2
96          6
97          5
98          1
99          1

[100 rows x 1 columns]

Jumlah data: 100


*(Kode untuk Langkah #4 ada di halaman 3 dokumen asli, akan ditempatkan di bawah ini)*

**Python Code (dari halaman 3, untuk Langkah #4 di sini): Menerapkan one-sample Z-test**

In [29]:
# Langkah 1: Mendefinisikan Hipotesis Nol dan Hipotesis Alternatif
# H0: Simulasi mungkin berasal dari dadu yang adil (rata-rata populasi = 3.5)
# H1: Simulasi kemungkinan tidak berasal dari dadu yang adil (rata-rata populasi != 3.5)

# Langkah 2 dan 3: Menghitung Nilai Z-Test dan p-Value
# Menggunakan data df_100 dan kolom 'lemparan:'
# Nilai yang diuji (rata-rata populasi H0) adalah 3.5
z_statistic_100, p_value_100 = ztest(df_100['lemparan:'], value=3.5)

# Print the results
print("--- Hasil Z-Test untuk 100 Lemparan (Simulasi Awal dari df_100) ---")
print(f"Z-Test statistic: {z_statistic_100}")
print(f"P-value: {p_value_100}")

# Langkah 4: Tarik Kesimpulan
alpha = 0.05 # Significance level
print(f"Alpha: {alpha}")
if p_value_100 < alpha:
    print("Kesimpulan: H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.")
else:
    print("Kesimpulan: H0 Diterima. Simulasi mungkin berasal dari dadu yang adil.")

--- Hasil Z-Test untuk 100 Lemparan (Simulasi Awal dari df_100) ---
Z-Test statistic: 1.4977836350257772
P-value: 0.13418947398894482
Alpha: 0.05
Kesimpulan: H0 Diterima. Simulasi mungkin berasal dari dadu yang adil.


**Bagaimana output dari Langkah 4 tersebut?**


--- Hasil Z-Test untuk 100 Lemparan (Simulasi Awal dari df_100) ---

- Z-Test statistic: 1.4977836350257772
- P-value: 0.13418947398894482
- Alpha: 0.05

Kesimpulan: H0 Diterima. Simulasi mungkin berasal dari dadu yang adil.

**Penjelasan Output:**
*   **Z-Test statistic:** Mengukur seberapa jauh rata-rata sampel dari rata-rata populasi yang dihipotesiskan (3.5), dalam satuan standar error. Karena dadu tidak adil (rata-rata sebenarnya > 3.5), kita mengharapkan Z positif.
*   **P-value:** Probabilitas mengamati data sampel seekstrem atau lebih ekstrem dari yang diamati, jika hipotesis nol (H0: rata-rata = 3.5) benar. P-value kecil menunjukkan bahwa data yang diamati tidak mungkin terjadi jika H0 benar.
*   **Alpha:** Tingkat signifikansi yang ditentukan (0.05 atau 5%).
*   **Kesimpulan:** H0 diterima (atau lebih tepatnya, gagal ditolak).

**Langkah #5: Simulasikan kembali data dan lakukan Z-Test kembali!**

In [30]:
# Langkah #5: Simulasi Ulang 1 (100 lemparan)
df_100_run2 = simulasi_dadu(100)
z_stat_100_run2, p_val_100_run2 = ztest(df_100_run2['lemparan:'], value=3.5)
kesimpulan_100_run2 = "H0 Ditolak" if p_val_100_run2 < alpha else "H0 Diterima"
detail_kesimpulan_100_run2 = f"Simulasi {'kemungkinan tidak' if p_val_100_run2 < alpha else 'mungkin'} berasal dari dadu yang adil."

print("--- Hasil Z-Test untuk 100 Lemparan (Simulasi Ulang 1) ---")
print(f"Z-Test = {z_stat_100_run2}")
print(f"p-Value = {p_val_100_run2}")
print(f"Kesimpulan : {kesimpulan_100_run2}. {detail_kesimpulan_100_run2}")

--- Hasil Z-Test untuk 100 Lemparan (Simulasi Ulang 1) ---
Z-Test = 1.436571796398436
p-Value = 0.15083970225879342
Kesimpulan : H0 Diterima. Simulasi mungkin berasal dari dadu yang adil.


|               | Hasil (Isi setelah menjalankan kode di atas)           |
|---------------|-------------------------------------------------------|
| **Z-Test =**  | *1.436571796398436*     |
| **p-Value =** | *0.15083970225879342*     |
| **Kesimpulan =**| *H0 Diterima. Simulasi mungkin berasal dari dadu yang adil.* |

**Langkah #6: Simulasikan kembali data dan lakukan Z-Test kembali!**

In [31]:
# Langkah #6: Simulasi Ulang 2 (100 lemparan)
df_100_run3 = simulasi_dadu(100)
z_stat_100_run3, p_val_100_run3 = ztest(df_100_run3['lemparan:'], value=3.5)
kesimpulan_100_run3 = "H0 Ditolak" if p_val_100_run3 < alpha else "H0 Diterima"
detail_kesimpulan_100_run3 = f"Simulasi {'kemungkinan tidak' if p_val_100_run3 < alpha else 'mungkin'} berasal dari dadu yang adil."

print("--- Hasil Z-Test untuk 100 Lemparan (Simulasi Ulang 2) ---")
print(f"Z-Test = {z_stat_100_run3}")
print(f"p-Value = {p_val_100_run3}")
print(f"Kesimpulan : {kesimpulan_100_run3}. {detail_kesimpulan_100_run3}")

--- Hasil Z-Test untuk 100 Lemparan (Simulasi Ulang 2) ---
Z-Test = 1.9148542155126769
p-Value = 0.05551109809787098
Kesimpulan : H0 Diterima. Simulasi mungkin berasal dari dadu yang adil.


|               | Hasil (Isi setelah menjalankan kode di atas)           |
|---------------|-------------------------------------------------------|
| **Z-Test =**  | *1.9148542155126769*     |
| **p-Value =** | *0.05551109809787098*     |
| **Kesimpulan :**| *H0 Diterima. Simulasi mungkin berasal dari dadu yang adil.* |

**Langkah #7: Simulasikan kembali data dan lakukan Z-Test kembali!**

In [32]:
# Langkah #7: Simulasi Ulang 3 (100 lemparan)
df_100_run4 = simulasi_dadu(100)
z_stat_100_run4, p_val_100_run4 = ztest(df_100_run4['lemparan:'], value=3.5)
kesimpulan_100_run4 = "H0 Ditolak" if p_val_100_run4 < alpha else "H0 Diterima"
detail_kesimpulan_100_run4 = f"Simulasi {'kemungkinan tidak' if p_val_100_run4 < alpha else 'mungkin'} berasal dari dadu yang adil."

print("--- Hasil Z-Test untuk 100 Lemparan (Simulasi Ulang 3) ---")
print(f"Z-Test = {z_stat_100_run4}")
print(f"p-Value = {p_val_100_run4}")
print(f"Kesimpulan : {kesimpulan_100_run4}. {detail_kesimpulan_100_run4}")

--- Hasil Z-Test untuk 100 Lemparan (Simulasi Ulang 3) ---
Z-Test = 2.6668462583203483
p-Value = 0.0076566688672452116
Kesimpulan : H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.


|               | Hasil (Isi setelah menjalankan kode di atas)           |
|---------------|-------------------------------------------------------|
| **Z-Test =**  | *2.6668462583203483*     |
| **p-Value =** | *0.0076566688672452116*     |
| **Kesimpulan :**| *H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.* |

**Langkah #8: Tuliskan kesimpulan anda dari beberapa kali simulasi yang anda lakukan!**

*(Isi berdasarkan hasil Langkah #4 (df_100), #5, #6, #7 untuk N=100).*

Dari beberapa kali simulasi dengan 100 lemparan dadu yang tidak adil (angka 6 muncul 2x lebih sering), dapat diamati bahwa:

1.  **Rata-rata Sampel:** Rata-rata sampel dari simulasi umumnya akan lebih tinggi dari 3.5 (rata-rata dadu adil). Rata-rata teoritis untuk dadu tidak adil ini adalah (1+2+3+4+5+6+6)/7 ≈ 3.857.
2.  **Nilai Z-Test:** Nilai Z-Test yang dihasilkan cenderung positif dan cukup besar. Ini menunjukkan bahwa rata-rata sampel secara signifikan berbeda (lebih tinggi) dari nilai hipotesis nol (3.5).
3.  **Nilai p-value:** Nilai p-value yang diperoleh dalam banyak kasus akan lebih kecil dari tingkat signifikansi alpha (0.05).
4.  **Kesimpulan Hipotesis:** Akibatnya, kesimpulan yang paling sering ditarik adalah "H0 Ditolak". Ini berarti ada cukup bukti statistik untuk menyatakan bahwa simulasi lemparan dadu tersebut kemungkinan besar tidak berasal dari dadu yang adil (yang memiliki rata-rata 3.5).
5.  **Variabilitas:** Meskipun demikian, karena adanya unsur acak dalam setiap simulasi, mungkin saja dalam beberapa run (meskipun kemungkinannya lebih kecil untuk N=100 dibandingkan N yang lebih kecil), p-value bisa saja lebih besar dari 0.05. Jika ini terjadi, kesimpulan akan menjadi "H0 Diterima" (gagal menolak H0), yang dalam konteks ini merupakan kesalahan Tipe II (gagal mendeteksi bahwa dadu tersebut memang tidak adil).

Secara umum, dengan N=100, Z-test seharusnya cukup sering dapat mendeteksi ketidakadilan dadu.

**Langkah #9: Untuk kondisi yang sama buat data simulasi lebih sedikit sebesar 50 data simulasi**

In [33]:
# Langkah #9: Simulasi 50 lemparan
df_50 = simulasi_dadu(50)
print("Data simulasi (50 lemparan) df_50:")
print(df_50.head())
print(f"\nBentuk DataFrame df_50: {df_50.shape}")

Data simulasi (50 lemparan) df_50:
   lemparan:
0          1
1          4
2          4
3          3
4          6

Bentuk DataFrame df_50: (50, 1)


**Langkah #10: Terapkan Z-Test dan tuliskan hasilnya!**

In [34]:
# Langkah #10: Z-Test untuk 50 lemparan (Simulasi Awal dari df_50)
z_stat_50, p_val_50 = ztest(df_50['lemparan:'], value=3.5)
kesimpulan_50 = "H0 Ditolak" if p_val_50 < alpha else "H0 Diterima"
detail_kesimpulan_50 = f"Simulasi {'kemungkinan tidak' if p_val_50 < alpha else 'mungkin'} berasal dari dadu yang adil."

print("--- Hasil Z-Test untuk 50 Lemparan (Simulasi Awal) ---")
print(f"Z-Test = {z_stat_50}")
print(f"p-Value = {p_val_50}")
print(f"Kesimpulan : {kesimpulan_50}. {detail_kesimpulan_50}")

--- Hasil Z-Test untuk 50 Lemparan (Simulasi Awal) ---
Z-Test = 3.4267223845497536
p-Value = 0.0006109132956922943
Kesimpulan : H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.


|               | Hasil (Isi setelah menjalankan kode di atas)           |
|---------------|-------------------------------------------------------|
| **Z-Test =**  | *3.4267223845497536*     |
| **p-Value =** | *0.0006109132956922943*     |
| **Kesimpulan :**| *H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.* |

**Langkah #11: Simulasikan kembali data dan lakukan Z-Test kembali! (N=50)**

In [35]:
# Langkah #11: Simulasi Ulang 1 (50 lemparan)
df_50_run2 = simulasi_dadu(50)
z_stat_50_run2, p_val_50_run2 = ztest(df_50_run2['lemparan:'], value=3.5)
kesimpulan_50_run2 = "H0 Ditolak" if p_val_50_run2 < alpha else "H0 Diterima"
detail_kesimpulan_50_run2 = f"Simulasi {'kemungkinan tidak' if p_val_50_run2 < alpha else 'mungkin'} berasal dari dadu yang adil."

print("--- Hasil Z-Test untuk 50 Lemparan (Simulasi Ulang 1) ---")
print(f"Z-Test = {z_stat_50_run2}")
print(f"p-Value = {p_val_50_run2}")
print(f"Kesimpulan : {kesimpulan_50_run2}. {detail_kesimpulan_50_run2}")

--- Hasil Z-Test untuk 50 Lemparan (Simulasi Ulang 1) ---
Z-Test = 2.7356321839080446
p-Value = 0.006226057504082859
Kesimpulan : H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.


|               | Hasil (Isi setelah menjalankan kode di atas)           |
|---------------|-------------------------------------------------------|
| **Z-Test =**  | *2.7356321839080446*     |
| **p-Value =** | *0.006226057504082859*     |
| **Kesimpulan :**| *H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.* |

**Langkah #12: Simulasikan kembali data dan lakukan Z-Test kembali! (N=50)**

In [36]:
# Langkah #12: Simulasi Ulang 2 (50 lemparan)
df_50_run3 = simulasi_dadu(50)
z_stat_50_run3, p_val_50_run3 = ztest(df_50_run3['lemparan:'], value=3.5)
kesimpulan_50_run3 = "H0 Ditolak" if p_val_50_run3 < alpha else "H0 Diterima"
detail_kesimpulan_50_run3 = f"Simulasi {'kemungkinan tidak' if p_val_50_run3 < alpha else 'mungkin'} berasal dari dadu yang adil."

print("--- Hasil Z-Test untuk 50 Lemparan (Simulasi Ulang 2) ---")
print(f"Z-Test = {z_stat_50_run3}")
print(f"p-Value = {p_val_50_run3}")
print(f"Kesimpulan : {kesimpulan_50_run3}. {detail_kesimpulan_50_run3}")

--- Hasil Z-Test untuk 50 Lemparan (Simulasi Ulang 2) ---
Z-Test = 1.4861379098733447
p-Value = 0.13724264667456407
Kesimpulan : H0 Diterima. Simulasi mungkin berasal dari dadu yang adil.


|               | Hasil (Isi setelah menjalankan kode di atas)           |
|---------------|-------------------------------------------------------|
| **Z-Test =**  | *1.4861379098733447*     |
| **p-Value =** | *0.13724264667456407*     |
| **Kesimpulan :**| *H0 Diterima. Simulasi mungkin berasal dari dadu yang adil.* |

**Langkah #13: Simulasikan kembali data dan lakukan Z-Test kembali! (N=50)**

In [37]:
# Langkah #13: Simulasi Ulang 3 (50 lemparan)
df_50_run4 = simulasi_dadu(50)
z_stat_50_run4, p_val_50_run4 = ztest(df_50_run4['lemparan:'], value=3.5)
kesimpulan_50_run4 = "H0 Ditolak" if p_val_50_run4 < alpha else "H0 Diterima"
detail_kesimpulan_50_run4 = f"Simulasi {'kemungkinan tidak' if p_val_50_run4 < alpha else 'mungkin'} berasal dari dadu yang adil."

print("--- Hasil Z-Test untuk 50 Lemparan (Simulasi Ulang 3) ---")
print(f"Z-Test = {z_stat_50_run4}")
print(f"p-Value = {p_val_50_run4}")
print(f"Kesimpulan : {kesimpulan_50_run4}. {detail_kesimpulan_50_run4}")

--- Hasil Z-Test untuk 50 Lemparan (Simulasi Ulang 3) ---
Z-Test = 2.753884535976096
p-Value = 0.005889254115493599
Kesimpulan : H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.


|               | Hasil (Isi setelah menjalankan kode di atas)           |
|---------------|-------------------------------------------------------|
| **Z-Test =**  | *2.753884535976096*     |
| **p-Value =** | *0.005889254115493599*     |
| **Kesimpulan :**| *H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.* |

**Langkah #14: Untuk kondisi yang sama buat data simulasi lebih sedikit sebesar 10.000 data simulasi**

*(Catatan: Terdapat kontradiksi "lebih sedikit" dan "10.000". Diasumsikan maksudnya adalah "lebih banyak" atau "sebesar 10.000" untuk menunjukkan efek sampel besar)*

In [38]:
# Langkah #14: Simulasi 10.000 lemparan
df_10000 = simulasi_dadu(10000)
print("Data simulasi (10.000 lemparan) df_10000:")
print(df_10000.head()) # Menampilkan sebagian kecil saja
print(f"\nBentuk DataFrame df_10000: {df_10000.shape}")

Data simulasi (10.000 lemparan) df_10000:
   lemparan:
0          5
1          5
2          5
3          6
4          5

Bentuk DataFrame df_10000: (10000, 1)


**Langkah #15: Terapkan Z-Test dan tuliskan hasilnya! (N=10000)**

In [39]:
# Langkah #15: Z-Test untuk 10.000 lemparan (Simulasi Awal dari df_10000)
z_stat_10000, p_val_10000 = ztest(df_10000['lemparan:'], value=3.5)
kesimpulan_10000 = "H0 Ditolak" if p_val_10000 < alpha else "H0 Diterima"
detail_kesimpulan_10000 = f"Simulasi {'kemungkinan tidak' if p_val_10000 < alpha else 'mungkin'} berasal dari dadu yang adil."

print("--- Hasil Z-Test untuk 10.000 Lemparan (Simulasi Awal) ---")
print(f"Z-Test = {z_stat_10000}")
print(f"p-Value = {p_val_10000}")
print(f"Kesimpulan : {kesimpulan_10000}. {detail_kesimpulan_10000}")

--- Hasil Z-Test untuk 10.000 Lemparan (Simulasi Awal) ---
Z-Test = 20.05996870761572
p-Value = 1.6518792909453674e-89
Kesimpulan : H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.


|               | Hasil (Isi setelah menjalankan kode di atas)           |
|---------------|-------------------------------------------------------|
| **Z-Test =**  | *20.05996870761572*     |
| **p-Value =** | *1.6518792909453674e-89*     |
| **Kesimpulan :**| *H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.* |

**Langkah #16: Simulasikan kembali data dan lakukan Z-Test kembali! (N=10000)**

In [40]:
# Langkah #16: Simulasi Ulang 1 (10.000 lemparan)
df_10000_run2 = simulasi_dadu(10000)
z_stat_10000_run2, p_val_10000_run2 = ztest(df_10000_run2['lemparan:'], value=3.5)
kesimpulan_10000_run2 = "H0 Ditolak" if p_val_10000_run2 < alpha else "H0 Diterima"
detail_kesimpulan_10000_run2 = f"Simulasi {'kemungkinan tidak' if p_val_10000_run2 < alpha else 'mungkin'} berasal dari dadu yang adil."

print("--- Hasil Z-Test untuk 10.000 Lemparan (Simulasi Ulang 1) ---")
print(f"Z-Test = {z_stat_10000_run2}")
print(f"p-Value = {p_val_10000_run2}")
print(f"Kesimpulan : {kesimpulan_10000_run2}. {detail_kesimpulan_10000_run2}")

--- Hasil Z-Test untuk 10.000 Lemparan (Simulasi Ulang 1) ---
Z-Test = 18.03636237403511
p-Value = 1.0097582366335608e-72
Kesimpulan : H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.


|               | Hasil (Isi setelah menjalankan kode di atas)           |
|---------------|-------------------------------------------------------|
| **Z-Test =**  | *18.03636237403511*     |
| **p-Value =** | *1.0097582366335608e-72*     |
| **Kesimpulan :**| *H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.* |

**Langkah #17: Simulasikan kembali data dan lakukan Z-Test kembali! (N=10000)**

In [41]:
# Langkah #17: Simulasi Ulang 2 (10.000 lemparan)
df_10000_run3 = simulasi_dadu(10000)
z_stat_10000_run3, p_val_10000_run3 = ztest(df_10000_run3['lemparan:'], value=3.5)
kesimpulan_10000_run3 = "H0 Ditolak" if p_val_10000_run3 < alpha else "H0 Diterima"
detail_kesimpulan_10000_run3 = f"Simulasi {'kemungkinan tidak' if p_val_10000_run3 < alpha else 'mungkin'} berasal dari dadu yang adil."

print("--- Hasil Z-Test untuk 10.000 Lemparan (Simulasi Ulang 2) ---")
print(f"Z-Test = {z_stat_10000_run3}")
print(f"p-Value = {p_val_10000_run3}")
print(f"Kesimpulan : {kesimpulan_10000_run3}. {detail_kesimpulan_10000_run3}")

--- Hasil Z-Test untuk 10.000 Lemparan (Simulasi Ulang 2) ---
Z-Test = 19.464447285512417
p-Value = 2.1985334796605344e-84
Kesimpulan : H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.


|               | Hasil (Isi setelah menjalankan kode di atas)           |
|---------------|-------------------------------------------------------|
| **Z-Test =**  | *19.464447285512417*     |
| **p-Value =** | *2.1985334796605344e-84*     |
| **Kesimpulan :**| *H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.* |

**Langkah #18: Simulasikan kembali data dan lakukan Z-Test kembali! (N=10000)**

In [42]:
# Langkah #18: Simulasi Ulang 3 (10.000 lemparan)
df_10000_run4 = simulasi_dadu(10000)
z_stat_10000_run4, p_val_10000_run4 = ztest(df_10000_run4['lemparan:'], value=3.5)
kesimpulan_10000_run4 = "H0 Ditolak" if p_val_10000_run4 < alpha else "H0 Diterima"
detail_kesimpulan_10000_run4 = f"Simulasi {'kemungkinan tidak' if p_val_10000_run4 < alpha else 'mungkin'} berasal dari dadu yang adil."

print("--- Hasil Z-Test untuk 10.000 Lemparan (Simulasi Ulang 3) ---")
print(f"Z-Test = {z_stat_10000_run4}")
print(f"p-Value = {p_val_10000_run4}")
print(f"Kesimpulan : {kesimpulan_10000_run4}. {detail_kesimpulan_10000_run4}")

--- Hasil Z-Test untuk 10.000 Lemparan (Simulasi Ulang 3) ---
Z-Test = 21.200569783469074
p-Value = 9.434731182662375e-100
Kesimpulan : H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.


|               | Hasil (Isi setelah menjalankan kode di atas)           |
|---------------|-------------------------------------------------------|
| **Z-Test =**  | *21.200569783469074*     |
| **p-Value =** | *9.434731182662375e-100*     |
| **Kesimpulan :**| *H0 Ditolak. Simulasi kemungkinan tidak berasal dari dadu yang adil.* |

**Langkah #19: Tuliskan kesimpulan anda dengan mengubah perubahan jumlah data simulasi dan setiap pengulangan proses simulasi !**

*(Isi berdasarkan hasil semua simulasi dengan N=50, N=100, dan N=10000, serta pengulangannya).*

Setelah melakukan serangkaian simulasi lemparan dadu tidak adil (angka 6 muncul 2x lebih sering) dengan berbagai ukuran sampel (N=50, N=100, N=10.000) dan melakukan pengulangan untuk setiap ukuran sampel, beberapa kesimpulan penting dapat ditarik mengenai pengaruh jumlah data dan pengulangan terhadap hasil Z-test:

1.  **Pengaruh Jumlah Data (Ukuran Sampel) terhadap Hasil Z-Test:**
    *   **N = 50 (Sampel Kecil):**
        *   Hasil Z-test menunjukkan variabilitas yang lebih tinggi. Rata-rata sampel bisa berfluktuasi cukup jauh dari rata-rata teoritis dadu tidak adil (≈3.857) hanya karena kebetulan acak.
        *   P-value juga lebih bervariasi. Ada kemungkinan yang lebih besar untuk mendapatkan p-value > 0.05, yang akan mengarah pada kesimpulan (keliru) bahwa H0 diterima (dadu dianggap adil). Ini mencerminkan *power* tes yang lebih rendah pada sampel kecil; tes kurang mampu mendeteksi perbedaan yang sebenarnya ada.
        *   Kesimpulan (Tolak/Terima H0) mungkin tidak konsisten antar pengulangan.
    *   **N = 100 (Sampel Sedang):**
        *   Variabilitas hasil Z-test dan p-value berkurang dibandingkan N=50. Rata-rata sampel cenderung lebih dekat ke ≈3.857.
        *   P-value lebih sering berada di bawah 0.05, sehingga H0 (dadu adil, rata-rata 3.5) lebih sering ditolak. *Power* tes meningkat.
        *   Kesimpulan menjadi lebih konsisten, dengan mayoritas pengulangan akan menolak H0.
    *   **N = 10.000 (Sampel Besar):**
        *   Dengan sampel yang sangat besar, rata-rata sampel akan sangat dekat dengan rata-rata teoritis dadu tidak adil (≈3.857). Standar error rata-rata menjadi sangat kecil.
        *   Nilai Z-test akan menjadi sangat besar (jauh dari nol), menunjukkan perbedaan yang sangat signifikan dari 3.5.
        *   P-value akan sangat kecil (mendekati nol), hampir selalu di bawah 0.05.
        *   Kesimpulan (Tolak H0) menjadi sangat konsisten di semua pengulangan. *Power* tes sangat tinggi; tes hampir pasti akan mendeteksi ketidakadilan dadu.

2.  **Pengaruh Pengulangan Proses Simulasi (untuk Ukuran Sampel yang Sama):**
    *   Pengulangan simulasi untuk ukuran sampel yang sama menunjukkan adanya **variabilitas sampling**. Setiap sampel acak, meskipun ditarik dari populasi yang sama dengan cara yang sama, akan sedikit berbeda.
    *   Akibatnya, nilai Z-statistik dan p-value akan sedikit berbeda di setiap pengulangan.
    *   Pada sampel kecil (N=50), variabilitas ini bisa cukup besar sehingga kesimpulan (Tolak/Terima H0) dapat berubah antar pengulangan.
    *   Pada sampel besar (N=10.000), meskipun Z dan p-value tetap berfluktuasi sedikit, fluktuasi tersebut biasanya tidak cukup besar untuk mengubah kesimpulan (H0 hampir selalu ditolak).

3.  **Kesimpulan Umum:**
    *   Dadu yang disimulasikan memang **tidak adil**. Rata-rata sebenarnya adalah ≈3.857, bukan 3.5.
    *   Z-test bertujuan untuk menguji hipotesis nol bahwa rata-rata populasi adalah 3.5. Semakin besar ukuran sampel, semakin akurat estimasi rata-rata sampel terhadap rata-rata populasi sebenarnya, dan semakin besar *power* Z-test untuk mendeteksi perbedaan antara rata-rata sebenarnya (≈3.857) dan rata-rata hipotesis nol (3.5).
    *   Eksperimen ini secara jelas menunjukkan bahwa **ukuran sampel sangat penting dalam pengujian hipotesis**. Sampel yang lebih besar memberikan bukti yang lebih kuat dan kesimpulan yang lebih andal. Dengan sampel kecil, kita berisiko lebih tinggi membuat kesalahan Tipe II (gagal menolak H0 yang salah).
    *   Pengulangan membantu memahami stabilitas hasil dan efek dari variabilitas sampling.